blog/2024-04-16-custom-components.md

```diff
@@ -0,0 +1,110 @@
+* Install third party components from PyPI with `pip`.
+* A new `rx.ComponentState` class that allows you to encapsulate state and UI for a component in a single class.
+
+## What are Custom Components?
+
+Reflex has many built-in components for common use cases, but sometimes you may want to create your own components. Custom components allow you to encapsulate complex logic and styling into a single reusable component.
+
+For example, in our [tutorial]({tutorial.final_app.path}) we share the code for creating a chat component in Reflex. In order to build a chat app, you would have to copy and paste the code for the chat component into your own app. 
+
+With custom components, we now have a [reflex-chat](https://github.com/picklelo/reflex-chat/) package that you can install with `pip` and use in your app.
+
+```bash
+pip install reflex-chat
+```
+
+You can then import the chat component into your app and use it like any other Reflex component. All the styling and logic is encapsulated, you only need to pass in the actual LLM logic.
+
+```python exec
+import reflex as rx
+from openai import OpenAI
+from reflex_chat import chat
+
+client = OpenAI()
+
+# Only define your logic, the chat component handles the rest.
+async def process(chat):
+    # Start a new session to answer the question.
+    session = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=chat.get_value(chat.messages),
+        stream=True,
+    )
+
+    # Stream the results, yielding after every word.
+    for item in session:
+        if hasattr(item.choices[0].delta, "content"):
+            answer_text = item.choices[0].delta.content
+            # Ensure answer_text is not None before concatenation
+            chat.messages[-1]["content"] += answer_text or ""
+            yield
+```
+
+```python demo
+rx.box(
+    chat(process=process),
+    height="500px",
+    width="100%"
+)
+```
+
+```md alert info
+# Just define your `process`function to connect the chat component to any LLM.
+
+\```python
+async def process(chat):
+    # Start a new session to answer the question.
+    session = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=chat.get_value(chat.messages),
+        stream=True,
+    )
+
+    # Stream the results, yielding after every word.
+    for item in session:
+        if hasattr(item.choices[0].delta, "content"):
+            answer_text = item.choices[0].delta.content
+            # Ensure answer_text is not None before concatenation
+            chat.messages[-1]["content"] += answer_text or ""
+            yield
+\```
+```
+
+## Types Of Custom Components
+
+There are two categories of custom components: **React wrappers** and **high-level components**.
+
+### React Wrapper
+
+Reflex is built on top of React, so if we don't have a component you need, you can easily [wrap your own React component]({wrapping_react.overview.path}) and use it in your Reflex app.
```

Comment:
```suggestion
Because Reflex is built on top of React, in most cases you can easily [wrap an existing React component]({wrapping_react.overview.path}) from NPM and use it in your Reflex app.
```